packages/styled-base/types/index.d.ts

```diff
@@ -33,50 +35,45 @@ export interface StyledComponent<InnerProps, StyleProps, Theme extends object>
   extends React.SFC<InnerProps & StyleProps & { theme?: Theme }>,
     ComponentSelector {
   /**
-   * @desc
-   * `Overwrapped` part checks the compatibility between `StyleProps` and `NewInnerProps`.
-   */
-  withComponent<NewInnerProps extends Overwrapped<StyleProps, NewInnerProps>>(
-    tag: React.ComponentType<NewInnerProps>
-  ): StyledComponent<NewInnerProps, StyleProps, Theme>
-
-  /**
-   * @desc this function overloading is unsafe
+   * @desc this method is type-unsafe
    */
-  withComponent<NewTag extends keyof JSX.IntrinsicElements>(
+  withComponent<NewTag extends keyof JSXInEl>(
     tag: NewTag
-  ): StyledComponent<JSX.IntrinsicElements[NewTag], StyleProps, Theme>
+  ): StyledComponent<JSXInEl[NewTag], StyleProps, Theme>
+  withComponent<Tag extends React.ComponentType<any>>(
+    tag: Tag
+  ): StyledComponent<PropsOf<Tag>, StyleProps, Theme>
```

Comment:
With this, we can use type of tag instead of `React.ComponentType` or something like that.
This is required for inferring valid type of ref.
(Originally, `ref` is `(el: React.ComponentType<Something>) => void`.)